ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java

```diff
@@ -567,35 +534,36 @@ private void readFromFile(File fileToRead, Journal journal) throws StoreExceptio
       short version = stream.readShort();
       switch (version) {
         case 0:
-          this.keySize = stream.readInt();
-          this.valueSize = stream.readInt();
+          keySize = stream.readInt();
+          valueSize = stream.readInt();
           long logEndOffset = stream.readLong();
           logger.trace("IndexSegment : {} reading log end offset {} from file", indexFile.getAbsolutePath(),
               logEndOffset);
           long maxEndOffset = Long.MIN_VALUE;
           while (stream.available() > Crc_Field_Length) {
             StoreKey key = factory.getStoreKey(stream);
-            byte[] value = new byte[IndexValue.Index_Value_Size_In_Bytes];
+            byte[] value = new byte[valueSize];
             stream.read(value);
-            IndexValue blobValue = new IndexValue(ByteBuffer.wrap(value));
+            IndexValue blobValue = new IndexValue(startOffset.getName(), ByteBuffer.wrap(value));
+            long offsetInLogSegment = blobValue.getOffset().getOffset();
             // ignore entries that have offsets outside the log end offset that this index represents
-            if (blobValue.getOffset() + blobValue.getSize() <= logEndOffset) {
+            if (offsetInLogSegment + blobValue.getSize() <= logEndOffset) {
               index.put(key, blobValue);
               logger.trace("IndexSegment : {} putting key {} in index offset {} size {}", indexFile.getAbsolutePath(),
                   key, blobValue.getOffset(), blobValue.getSize());
               // regenerate the bloom filter for in memory indexes
               bloomFilter.add(ByteBuffer.wrap(key.toBytes()));
               // add to the journal
-              if (blobValue.getOffset() != blobValue.getOriginalMessageOffset()
-                  && blobValue.getOriginalMessageOffset() >= startOffset.get()) {
+              if (offsetInLogSegment != blobValue.getOriginalMessageOffset()
+                  && blobValue.getOriginalMessageOffset() >= startOffset.getOffset()) {
                 // we add an entry for the original message offset if it is within the same index segment
-                journal.addEntry(blobValue.getOriginalMessageOffset(), key);
+                journal.addEntry(new Offset(startOffset.getName(), blobValue.getOriginalMessageOffset()), key);
               }
               journal.addEntry(blobValue.getOffset(), key);
-              sizeWritten.addAndGet(key.sizeInBytes() + IndexValue.Index_Value_Size_In_Bytes);
+              sizeWritten.addAndGet(key.sizeInBytes() + valueSize);
               numberOfItems.incrementAndGet();
-              if (blobValue.getOffset() + blobValue.getSize() > maxEndOffset) {
-                maxEndOffset = blobValue.getOffset() + blobValue.getSize();
+              if (offsetInLogSegment + blobValue.getSize() > maxEndOffset) {
+                maxEndOffset = offsetInLogSegment + blobValue.getSize();
               }
             } else {
               logger.info(
```

Comment:
Given how `writeToFile()` is done, I think the `else` case here will never happen unless there is a programming error - and if it happens it should be logged as an error and not info. Do you think there are valid cases when this can happen?